File: shared/definitions/changes.py
"""What is new to hunt: surface the world added, never what the operator did."""

from __future__ import annotations

from datetime import timedelta
from enum import StrEnum

from shared.definitions.bounty_programs import BountyEvent


class ChangeKind(StrEnum):
    SCANNED_ASSET = "scanned_asset"
    SCOPE_ASSET = "scope_asset"
    PROGRAM = "program"
    WATCH_HOST = "watch_host"
    TARGET = "target"


KIND_ORDER: tuple[str, ...] = tuple(k.value for k in ChangeKind)

KIND_LABELS: dict[str, str] = {
    ChangeKind.SCANNED_ASSET.value: "New scanned asset",
    ChangeKind.SCOPE_ASSET.value: "New in-scope asset",
    ChangeKind.PROGRAM.value: "New program",
    ChangeKind.WATCH_HOST.value: "New watched host",
    ChangeKind.TARGET.value: "New target",
}

BOUNTY_KINDS: frozenset[str] = frozenset(
    {
        ChangeKind.SCOPE_ASSET.value,
        ChangeKind.PROGRAM.value,
        ChangeKind.WATCH_HOST.value,
        ChangeKind.TARGET.value,
    }
)


class ChangeBasis(StrEnum):
    MARK = "mark"
    WINDOW = "window"


class ChangeTone(StrEnum):
    NEW = "new"
    HOT = "hot"
    NEUTRAL = "neutral"


SOURCE_LABELS: dict[str, str] = {
    "ct_log": "Certificate log",
    "watch": "Program watch",
    "scope": "Program scope",
    "api": "Platform API",
    "feed": "Public feed",
}

CHANGE_WINDOWS: dict[str, timedelta] = {
    "7d": timedelta(days=7),
    "30d": timedelta(days=30),
    "90d": timedelta(days=90),
}
DEFAULT_CHANGE_WINDOW = "30d"
CHANGE_FEED_LIMIT = 300

EVENT_TONES: dict[str, str] = {
    "info": ChangeTone.NEW.value,
    "warning": ChangeTone.HOT.value,
    "muted": ChangeTone.NEUTRAL.value,
}

SCOPE_EVENTS: frozenset[str] = frozenset(
    {BountyEvent.SCOPE_ADDED.value, BountyEvent.CAME_INTO_SCOPE.value}
)
PROGRAM_EVENTS: frozenset[str] = frozenset(
    {
        BountyEvent.PROGRAM_ADDED.value,
        BountyEvent.PROGRAM_WENT_PUBLIC.value,
        BountyEvent.SUBMISSIONS_OPENED.value,
        BountyEvent.BOUNTIES_STARTED.value,
    }
)
NEW_EVENTS: frozenset[str] = SCOPE_EVENTS | PROGRAM_EVENTS


def event_kinds_for(kinds: set[str]) -> set[str]:
    out: set[str] = set()
    if ChangeKind.SCOPE_ASSET.value in kinds:
        out |= SCOPE_EVENTS
    if ChangeKind.PROGRAM.value in kinds:
        out |= PROGRAM_EVENTS
    return out


def change_kind_of_event(kind: str) -> str | None:
    if kind in SCOPE_EVENTS:
        return ChangeKind.SCOPE_ASSET.value
    if kind in PROGRAM_EVENTS:
        return ChangeKind.PROGRAM.value
    return None


def mark_key(project_id) -> str:
    return f"changes:{project_id}"

File: frontend/src/lib/config/changes.ts
// mirrors shared/definitions/changes.py
import Globe from '@lucide/svelte/icons/globe';
import ShieldCheck from '@lucide/svelte/icons/shield-check';
import Award from '@lucide/svelte/icons/award';
import Target from '@lucide/svelte/icons/target';
import Radar from '@lucide/svelte/icons/radar';
import type { IconComponent } from './icons';

export const ChangeKind = {
	SCANNED_ASSET: 'scanned_asset',
	SCOPE_ASSET: 'scope_asset',
	PROGRAM: 'program',
	WATCH_HOST: 'watch_host',
	TARGET: 'target'
} as const;
export type ChangeKindKey = (typeof ChangeKind)[keyof typeof ChangeKind];

export const KIND_ORDER: ChangeKindKey[] = [
	ChangeKind.SCANNED_ASSET,
	ChangeKind.SCOPE_ASSET,
	ChangeKind.PROGRAM,
	ChangeKind.WATCH_HOST,
	ChangeKind.TARGET
];

export const KIND_LABELS: Record<ChangeKindKey, string> = {
	[ChangeKind.SCANNED_ASSET]: 'Scanned assets',
	[ChangeKind.SCOPE_ASSET]: 'In-scope assets',
	[ChangeKind.PROGRAM]: 'Programs',
	[ChangeKind.WATCH_HOST]: 'Watched hosts',
	[ChangeKind.TARGET]: 'Targets'
};

export const KIND_ICONS: Record<ChangeKindKey, IconComponent> = {
	[ChangeKind.SCANNED_ASSET]: Globe,
	[ChangeKind.SCOPE_ASSET]: ShieldCheck,
	[ChangeKind.PROGRAM]: Award,
	[ChangeKind.WATCH_HOST]: Radar,
	[ChangeKind.TARGET]: Target
};

export const BOUNTY_KINDS: ReadonlySet<string> = new Set([
	ChangeKind.SCOPE_ASSET,
	ChangeKind.PROGRAM,
	ChangeKind.WATCH_HOST,
	ChangeKind.TARGET
]);

export const ChangeBasis = { MARK: 'mark', WINDOW: 'window' } as const;

export const CHANGE_MODES = [
	{ key: 'since', label: 'Since last visit' },
	{ key: 'timeline', label: 'Timeline' }
] as const;
export type ChangeMode = (typeof CHANGE_MODES)[number]['key'];

export const CHANGE_WINDOWS = [
	{ key: '7d', label: '7d', text: 'last 7 days' },
	{ key: '30d', label: '30d', text: 'last 30 days' },
	{ key: '90d', label: '90d', text: 'last 90 days' }
] as const;
export type ChangeWindow = (typeof CHANGE_WINDOWS)[number]['key'];
export const DEFAULT_CHANGE_WINDOW: ChangeWindow = '30d';

export const TONE_NODE: Record<string, string> = {
	new: 'bg-info/10 text-info ring-1 ring-info/30',
	hot: 'bg-destructive/10 text-destructive ring-1 ring-destructive/40',
	neutral: 'bg-muted text-muted-foreground ring-1 ring-border'
};

export const GRID_STEPS = 4;
